tests/unit/s2n_self_talk_inet_socket_test.c

```diff
@@ -0,0 +1,177 @@
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+#define S2N_TEST_INADDR_LOOPBACK 0x7f000001 /* 127.0.0.1 */
+
+#define WRITE_sync()                                             \
+    EXPECT_SUCCESS(sleep(1)); /* allow time for data to flush */ \
+    EXPECT_SUCCESS(write(write_pipe, &sync, 1));                 \
+    EXPECT_SUCCESS(sleep(1)); /* allow reader time time process current operation */
+
+#define READ_sync() \
+    EXPECT_SUCCESS(read(read_pipe, &sync, 1));
+
+#define SEND_CHAR(conn, c) \
+    send_buffer[0] = c;    \
+    EXPECT_SUCCESS(s2n_send(conn, send_buffer, 1, &blocked));
+
+const char a = 'a';
+const char b = 'b';
+
+static S2N_RESULT start_client(int fd, int read_pipe)
+{
+    /* Setup connections */
+    DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+            s2n_connection_ptr_free);
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key * chain_and_key, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Setup config */
+    EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+    EXPECT_SUCCESS(s2n_connection_set_fd(client_conn, fd));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+    EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+    /* Do handshake */
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    EXPECT_SUCCESS(s2n_negotiate(client_conn, &blocked));
+    EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS12);
+
+    /* Set connection to non-blocking mode */
+    int flags = fcntl(fd, F_GETFL, 0);
+    EXPECT_SUCCESS(fcntl(fd, F_SETFL, flags | O_NONBLOCK));
+
+    char sync;
+    char recv_buffer[10];
+
+    {
+        READ_sync();
+        EXPECT_SUCCESS(s2n_recv(client_conn, recv_buffer, 1, &blocked));
+        EXPECT_TRUE(memcmp(&a, &recv_buffer[0], 1) == 0);
+
+        READ_sync();
+        EXPECT_SUCCESS(s2n_recv(client_conn, recv_buffer, 1, &blocked));
+        EXPECT_TRUE(memcmp(&b, &recv_buffer[0], 1) == 0);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT start_server(int fd, int write_pipe)
+{
+    /* Setup connections */
+    DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+            s2n_connection_ptr_free);
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key * chain_and_key, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
```

Comment:
If you're returning S2N_RESULT, why not use the normal safety macros instead of the `EXPECT`?